### PR TITLE
Improve performance of admin/conference/*/registrations

### PR DESCRIPTION
**Pull request by bear454**
* Add some caching
* Reduce query count, including N+1s on questions & roles

**Checklist**

- [x] I have read the [Contribution & Best practices Guide](https://github.com/openSUSE/osem/blob/master/CONTRIBUTING.md).
- [x] My branch is up-to-date with the upstream `master` branch.
- [x] The tests pass locally with my changes.

**Short description of what this resolves/which [issues](https://github.com/openSUSE/osem/issues) does this fix?:**

- Registrations page was so slow it wouldn't load anymore (Total render time: 123 seconds)

**Changes proposed in this pull request:**

- Add some caching
- Reduce query count, including N+1s on questions & roles

Total render time: 6 seconds. Still slow, but, you know, 1/20 of the time before ;-)

### CODE REVIEW

File: app/controllers/admin/registrations_controller.rb [hunk 1]
```diff
@@ -6,12 +6,17 @@ class RegistrationsController < Admin::BaseController
 
     def index
       authorize! :show, Registration.new(conference_id: @conference.id)
-      @pdf_filename = "#{@conference.title}.pdf"
-      @registrations = @conference.registrations.includes(:user).order('registrations.created_at ASC')
-      @attended = @conference.registrations.where('attended = ?', true).count
+      @registrations = @conference.registrations.eager_load(
```
Comment (Ana06) [inline app/controllers/admin/registrations_controller.rb]: why `eager_load` instead of `includes`? :thinking:

File: app/controllers/admin/registrations_controller.rb [hunk 2]
```diff
@@ -6,12 +6,17 @@ class RegistrationsController < Admin::BaseController
...
+        :qanswers, user: :roles
+      ).order('registrations.created_at ASC').to_a
```
Comment (Ana06) [inline app/controllers/admin/registrations_controller.rb]: why is it needed to convert it to array? :confused:

File: app/views/admin/registrations/_questions.html.haml [hunk 3]
```diff
@@ -1,9 +1,5 @@
...
-
-  %b Q:
-  = q.title
-
-  %b A:
-  - registration.qanswers.where(question_id: q.id).each do |qa|
-    = qa.answer.title
-  %br
+- cache [:admin, @conference, registration.qanswers] do
+  %dl
```
Comment (Ana06) [inline app/views/admin/registrations/_questions.html.haml]: why are you changing `%b` by `%dl`? how is that related to the performance? :confused:

File: app/controllers/admin/registrations_controller.rb [hunk 4]
```diff
@@ -6,12 +6,17 @@ class RegistrationsController < Admin::BaseController
...
+      @attended = @registrations.count(&:attended)
+      @questions = @conference.questions.to_a
```
Comment (Ana06) [inline app/controllers/admin/registrations_controller.rb]: why is this variable needed? :thinking:

File: app/views/admin/registrations/index.pdf.prawn [hunk 5]
```diff
@@ -23,10 +23,8 @@ prawn_document(force_download: true, filename: @pdf_filename, page_layout: :land
     row << registration.departure.to_s || ''
 
     @conference.questions.each do |question|
-      qa = registration.qanswers.find_by(question: question)
-      answer = ( qa ? qa.answer.title : '' )
-
-      row << answer
+      qa = registration.qanswers.find{|qa| qa.question_id == question.id }
```
Comment (Ana06) [inline app/views/admin/registrations/index.pdf.prawn]: why is that better than `qa = registration.qanswers.find_by(question: question)`? :thinking:

File: app/views/admin/registrations/index.xlsx.axlsx [hunk 6]
```diff
@@ -3,7 +3,7 @@ wb.add_worksheet(name: 'registrations') do |sheet|
   bold_style = wb.styles.add_style(b: true)
   row = ['Attended', 'Name', 'Nickname', 'Affilιation', 'Email', 'Arrival', 'Departure']
 
-  @conference.questions.each do |question|
+  @conference.questions.sort_by(&:id).each do |question|
```
Comment (Ana06) [inline app/views/admin/registrations/index.xlsx.axlsx]: This is not related with the performance, right? I'd move it to a different commit

File: app/controllers/admin/registrations_controller.rb (hunk 1 above)
Comment (bear454) [reply app/controllers/admin/registrations_controller.rb]: `eager_load` instantiates the dependent object and does so in a _single_ query; `includes` adds additional queries for each included model.

http://guides.rubyonrails.org/active_record_querying.html#eager-loading-associations

File: app/controllers/admin/registrations_controller.rb (hunk 2 above)
Comment (bear454) [reply app/controllers/admin/registrations_controller.rb]: ActiveRecord emulates many of the functions in Enumerable, but performs additional queries when doing so; e.g. ActiveRelation#count performs a "SELECT count(*) FROM" whereas Enumerable#count is performed on the loaded collection without another roundtrip to the database.

File: app/views/admin/registrations/_questions.html.haml (hunk 3 above)
Comment (bear454) [reply app/views/admin/registrations/_questions.html.haml]: The Question partial had to be redesigned to accommodate the different data model.

File: app/views/admin/registrations/index.pdf.prawn (hunk 5 above)
Comment (bear454) [reply app/views/admin/registrations/index.pdf.prawn]: `ActiveRelation#find_by` peforms an additonal query against the database, for something that can be solved much faster in Ruby.

File: app/controllers/admin/registrations_controller.rb (hunk 4 above)
Comment (bear454) [reply app/controllers/admin/registrations_controller.rb]: Readability.

File: app/views/admin/registrations/index.xlsx.axlsx (hunk 6 above)
Comment (bear454) [reply app/views/admin/registrations/index.xlsx.axlsx]: In order to improve parity with the old design, the questions needed to be sorted consistently.

**Comment by bear454** [comment]
@Ana06 It's been > 3 weeks since last comments. Do you have any further issues, or should I consider your comments resolved?

File: app/views/admin/registrations/index.pdf.prawn (hunk 5 above)
Comment (Ana06) [reply app/views/admin/registrations/index.pdf.prawn]: database queries are normally much faster than manipulating Ruby objects... :confused: 

`registration.qanswers.find{|qa| qa.question_id == question.id }` will also triger a query to the database

```
> registration = Registration.last
  Registration Load (0.1ms)  SELECT  "registrations".* FROM "registrations" ORDER BY "registrations"."id" DESC LIMIT ?  [["LIMIT", 1]]
=> #<Registration id: 14433, conference_id: 1, arrival: nil, departure: nil, created_at: "2018-06-15 12:57:31", updated_at: "2018-06-15 12:57:31", other_special_needs: nil, attended: false, volunteer: nil, user_id: 14437, week: 24, accepted_code_of_conduct: nil>

> registration.qanswers.find{|qa| qa.question_id == 2 }
  Qanswer Load (0.1ms)  SELECT "qanswers".* FROM "qanswers" INNER JOIN "qanswers_registrations" ON "qanswers"."id" = "qanswers_registrations"."qanswer_id" WHERE "qanswers_registrations"."registration_id" = ?  [["registration_id", 14433]]

> registration.qanswers.find(2)
  Qanswer Load (6.7ms)  SELECT  "qanswers".* FROM "qanswers" INNER JOIN "qanswers_registrations" ON "qanswers"."id" = "qanswers_registrations"."qanswer_id" WHERE "qanswers_registrations"."registration_id" = ? AND "qanswers"."id" = ? LIMIT ?  [["registration_id", 14433], ["id", 2], ["LIMIT", 1]]
```

Comment (bear454) [reply app/views/admin/registrations/index.pdf.prawn]: That depends on latency to your database. Osem is loaded with inefficient queries that are masked by high database response rate.

For LFNW, our database is disparate from our application server; the latency "lowers the water" to show the rocks that impede flow.

Strategies like this have been employed on other problem areas, especially the splashpage, resulting in > 30x performance improvement overall; that performance benefits everyone (even if you don't see it in your query response time) by generating less waste and utilizing limited resources more efficiently.

File: app/controllers/admin/registrations_controller.rb (hunk 2 above)
Comment (bear454) [reply app/controllers/admin/registrations_controller.rb]: `@registrations` is redefined her for perfomance; I think you're confusing it with `@attended` which is defined on the next line :arrow_down: as `@registrations.count(&:attended)`.

File: app/controllers/admin/registrations_controller.rb [hunk 7]
```diff
@@ -8,13 +8,18 @@ class RegistrationsController < Admin::BaseController
...
+      @roles = @conference.roles.to_a
```
[the remark this answers is not in the document]
Comment (bear454) [reply app/controllers/admin/registrations_controller.rb]: It looks like not ;-) ... I'll remove it.

File: app/controllers/admin/registrations_controller.rb (hunk 4 above)
Comment (bear454) [reply app/controllers/admin/registrations_controller.rb]: When we iterate over the questions, it caused ActiveRecord's enumberable overrides to perform additional (unnecessary) queries.

File: app/controllers/admin/registrations_controller.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (bear454) [reply app/controllers/admin/registrations_controller.rb]: Removed in 9d4e559c.

**Comment by hennevogel** [comment]
@bear454 @Ana06 are we going to get this in?

**Comment by bear454** [comment]
@hennevogel I'm closing it; Moving to ajax-datatables will make most of the improvements moot, anyway. Thanks for the reminder.
